pkg/infrastructure/aws/clusterapi/aws.go

```diff
@@ -247,15 +261,47 @@ func (p *Provider) DestroyBootstrap(ctx context.Context, in clusterapi.Bootstrap
 		return fmt.Errorf("failed to create aws session: %w", err)
 	}
 
-	if err := removeSSHRule(ctx, in.Client, in.Metadata.InfraID, region, session, awsCluster); err != nil {
-		return fmt.Errorf("failed to remove bootstrap SSH rule: %w", err)
+	timeout := 15 * time.Minute
+	startTime := time.Now()
+	untilTime := startTime.Add(timeout)
+	timezone, _ := untilTime.Zone()
+	logrus.Debugf("Waiting up to %v (until %v %s) for bootstrap SSH rule to be destroyed...", timeout, untilTime.Format(time.Kitchen), timezone)
+	if err := wait.PollUntilContextTimeout(ctx, 15*time.Second, timeout, true,
+		func(ctx context.Context) (bool, error) {
+			if err := removeSSHRule(ctx, in.Client, in.Metadata.InfraID); err != nil {
+				// If the cluster object has been modified between Get and Update, k8s client will refuse to update it.
+				// In that case, we need to retry.
+				if k8serrors.IsConflict(err) {
+					logrus.Debugf("AWSCluster update conflict during SSH rule removal: %v", err)
+					return false, nil
+				}
+				return true, fmt.Errorf("failed to remove bootstrap SSH rule: %w", err)
+			}
+			return isSSHRuleGone(ctx, session, region, sgID)
```

Comment:
This is called immediately after the `removeSSHRule`. Does there need to be a "wait" or confirmation of removal operation just to make sure that the operation completed (it is asynchronous I believe)?

Reply:
I don't think it's needed. It'll not be removed the first time, but then it might be in the second iteration.